### ISSUE TITLE
array.overlap and array.map_overlap block sizes are incorrect when depth is an unsigned bit type

### ISSUE THREAD

**Issue by GFleishman** [dask/dask#6849]
**Minimal Example**:
```python
import numpy as np
import dask.array as da

def func(block, block_info=None):
    print(block_info[0]['array-location'])
    print(block.shape)
    return block

x = da.ones((100,), chunks=(50,50))

d_signed = tuple( np.array([10, 10]).astype(np.int16) )
y_signed = da.map_overlap(func, x, dtype=x.dtype, depth=d_signed)

d_unsigned = tuple( np.array([10, 10]).astype(np.uint16) )
y_unsigned = da.map_overlap(func, x, dtype=x.dtype, depth=d_unsigned)
```

```python
y_signed.compute()
```
[(0, 70)]
[(70, 140)]
(70,)
(70,)

```python
y_unsigned.compute()
```
[(70, 110)]
(40,)
[(0, 70)]
(60,)

I gather that `map_overlap` simply pads each block with data from its neighbors, then calls `map_blocks` on the "augmented" array. I suspect that when this augmented array is sliced there is probably an expression like:
`new_block = augmented_array[d[0]:-d[0], ...]` where `d` is the depth tuple. Simply taking the negative for the end of the slice does not work for unsigned bit types. This should be replaced with `augmented_array[d[0]:-1*d[0], ...]` to ensure any unsigned bit type is properly cast before slicing.

Why should we support unsigned bit type arguments here?

The depth argument is intrinsically an unsigned quantity, there is no "negative depth" in this context. It's reasonable for users to expect this data type to be accepted. When depths are computed automatically from things like array shapes (e.g. `my_depth = tuple( np.array( my_array.shape // 8 ) )` for an approximate 12.5% overlap) they will often be unsigned bit types.

**Environment**:

- Dask version: 2.20.0
- Python version: 3.6.4
- Operating System: Scientific Linus
- Install method (conda, pip, source): pip

**Comment by jsignell** [comment]
That argument makes sense to me and I can reproduce on latest master. It seems like you have a good handle on how to implement the fix, so please feel free to open a Pull Request.

**Comment by GFleishman** [comment]
Thanks for reproducing, I'm happy to take a look at it this evening :)

**Comment by GFleishman** [comment]
Sorry for delay - I'm still willing to fix and submit PR, just have deliverables at work that take priority. I may take some time over the holiday to look into this.

**Comment by GFleishman** [comment]
Decided to look into it now after all.

The bug is nested down to [`da.overlap.overlap`](https://github.com/dask/dask/blob/2ce9cf9190482315e317656e11cd1f7707f91c0c/dask/array/overlap.py#L472) (try running previous example but use `da.overlap.overlap` instead of `map_overlap`). It likely nests further down to [`da.overlap.overlap_internal`](https://github.com/dask/dask/blob/2ce9cf9190482315e317656e11cd1f7707f91c0c/dask/array/overlap.py#L119) and potentially further than that (have not looked yet).

Both `da.map_overlap` and `da.overlap.overlap` call [`da.coerce_depth`](https://github.com/dask/dask/blob/2ce9cf9190482315e317656e11cd1f7707f91c0c/dask/array/overlap.py#L811) to reformat the depth data structure. The simplest solution would be to wrap the depth elements with `int()` here in this parsing function.

Alternatively, I could follow the nested trail of overlap functions and find where array slicing/padding actually happens and cast to `int` there.

I believe either is fine and that it's really a matter of design/cohesion. What is the prevailing philosophy for this kind of thing in dask? Where are inputs standardized? Is modifying `coerce_depth` a kludge or acceptable?

**Comment by jsignell** [comment]
> Both da.map_overlap and da.overlap.overlap call da.coerce_depth to reformat the depth data structure. The simplest solution would be to wrap the depth elements with int() here in this parsing function.

This seems like a reasonable solution to me. Type coersion should happen sooner rather than later so that it can fail quickly.

**Comment by GFleishman** [comment]
Forked and fixed on a new branch. Submitting PR, it asks for "- [ ] Tests added / passed- [ ] Passes `black dask` / `flake8 dask`" Can you share a link to documentation on these requirements? I'm happy to meet them, but not sure what they are.

**Comment by jsignell** [comment]
The place to put tests would be: https://github.com/dask/dask/blob/master/dask/array/tests/test_overlap.py 

And this section covers black and flake8: https://docs.dask.org/en/latest/develop.html#code-formatting

**Comment by GFleishman** [comment]
Ahh, sorry I missed the dask/array/tests folder - just an oversight. Thanks!

**Comment by GFleishman** [comment]
PR submitted: https://github.com/dask/dask/pull/6909
Thanks for the guidance @jsignell !

**Comment by jsignell** [comment]
Closed by #6909! Thanks @GFleishman
